script/hassfest/quality_scale_validation/runtime_data.py

```diff
@@ -0,0 +1,44 @@
+"""Enforce that the integration only uses ConfigEntry.runtime_data."""
+
+import ast
+from collections.abc import Generator
+import pathlib
+
+from script.hassfest.model import Integration
+
+
+def _integration_python_files(
+    integration: Integration,
+) -> Generator[pathlib.Path, None, None]:
+    """Return all python files in the integration."""
+    for root, _dirs, files in integration.path.walk():
+        for file in files:
+            if file.endswith(".py"):
+                yield root / file
+
+
+def _has_hass_data(tree: ast.Module) -> bool:
+    """Test if the module uses hass.data."""
+    for node in ast.walk(tree):
+        if (
+            isinstance(node, ast.Attribute)
+            and isinstance(node.value, ast.Name)
+            and node.value.id == "hass"
+            and node.attr == "data"
+        ):
+            return True
+    return False
+
+
+def validate(integration: Integration) -> list[str] | None:
+    """Validate that the integration does not use hass.data."""
+    rules = []
+    for file_path in _integration_python_files(integration):
+        module = ast.parse(file_path.read_text())
+        if _has_hass_data(module):
+            rules.append(
+                "Integration is not exclusively using ConfigEntry.runtime_data "
+                f"(found use of hass.data in {file_path}). You may add an exemption "
+                "to this using hass.data for global storage not specific to a config entry.)",
+            )
```

Comment:
Note that the rule is not to exclusively use runtime data, there are integrations like iron_os and wled that have a shared coordinator stored in hass.data

Reply:
Yes this is essentially proposing to make the rule about avoiding use of hass.data, but allowing it in exceptional cases when needed. 

Would you prefer we do not do this check at all or have a different suggestion?

Reply:
Can we check the type of the second parameter of `async_setup_entry`? If that is ConfigEntry, we most likely did not implement entry.runtime_data, or we are missing typing

Reply:
Makes sense. One challenge there is there may not be any need to use runtime_data (which is why i think the rule is worded slightly backward, as I interpreted it more about banning use of hass.data for config entry runtime data rather than forcing config_entry runtime data always used).  Would you say an integration should be exempt in that case? If so, then I can give that a try.

Other variations I've considered:
- Look at the parameters to hass.data ... and see if its a ConfigEntry.entry_id. this seemed really hard to do.
- Don't allow setting `hass.data` while inside of the function `async_setup_entry`. I was able to implement something like this with a visitor, so that could work.

Reply:
I think the phrasing for exemptions should be moved out of the validator into quality_scale.py

```python
if errors:
    # loop to add integration errors
    # then add note about exemptions and justification 
```

Reply:
> Don't allow setting hass.data while inside of the function async_setup_entry. I was able to implement something like this with a visitor, so that could work.

This would probably work for the use cases that iron_os and wled have, since those are setup in `async_setup`. I can't think of any use case other than that.

One integration that we also need to keep in mind is MQTT. Even though it is marked as single_config_entry. We still use hass.data, I believe we do that for other integrations to access the broker.

> Would you say an integration should be exempt in that case? If so, then I can give that a try.

I think if an integration does not have runtime data, it would be exempted

Reply:
> Can we check the type of the second parameter of `async_setup_entry`? If that is ConfigEntry, we most likely did not implement entry.runtime_data, or we are missing typing

The `twentemilieu` currently uses `runtime_data` without any type change. I'm assuming that is Ok because strict typing is required in a later level, but we can check.

I've updated this to check for `ConfigEntry.runtime_data` inside of `async_setup_entry` so its more literally enforcing the rule. This won't work if we set it in another function, but maybe that is ok? (I noticed a similar problem checking for service registration when its done in services.py)

Reply:
I do believe that implementing runtime_data was done for two reasons:
- add type information
- AND avoid polluting/storing information in hass.data

Should there be a separate rule for avoiding direct access to `hass.data` (unless exempt because of shared content)?
And in that case, hass.data should be accessed via HassKey not via direct string.

Reply:
After careful consideration, I think we can add `hass.data` checks in a follow-up PR